### PR TITLE
GO-3040 add ft fields to the objectSearch responses

### PR DESCRIPTION
**Pull request by requilence**
- Add ObjectSearchWithMeta RPC
   - records now wrapped into SearchResult
   - each result contains: Highlight, HighlightRanges, BlockId, RelationKey, RelationDetails
- Rewrite fulltext flow
   - pass the whole fulltext results to the next step instead of list of IDs
   -  add the "_score" to the result

### CODE REVIEW

**Review by fat-fellow — CHANGES_REQUESTED** [review]
Can we have some testing here?

File: pkg/lib/localstore/objectstore/queries.go [hunk 1]
```diff
@@ -43,7 +44,19 @@ func (s *dsObjectStore) QueryRaw(filters *database.Filters, limit int, offset in
 				return err
 			}
 			rec := database.Record{Details: details.Details}
-			// todo: pass the inner block/relation to the result
+			id := pbtypes.GetString(details.Details, bundle.RelationKeyId.String())
```
Comment (fat-fellow) [inline pkg/lib/localstore/objectstore/queries.go:47]: maybe here

File: pkg/lib/localstore/objectstore/queries.go [hunk 2]
```diff
@@ -125,9 +138,24 @@ func (s *dsObjectStore) makeFTSQuery(text string, filters *database.Filters) (*d
 		return objectResults[i].Score > objectResults[j].Score
 	})
 
+	filters.FulltextMetadata = make(map[string]database.FulltextMetadata)
+
 	var objectIds = make([]string, 0, len(objectResults))
 	for _, result := range objectResults {
-		objectIds = append(objectIds, domain.NewFromPath(result.ID).ObjectId)
+		path := domain.NewFromPath(result.ID)
+		var highlight string
```
Comment (fat-fellow) [inline pkg/lib/localstore/objectstore/queries.go:146]: maybe here

File: pkg/lib/localstore/objectstore/queries_test.go [hunk 3]
```diff
@@ -21,6 +21,20 @@ import (
...
+	}
+	return records
+}
+
+func removeMetaFromRecords(records []database.Record) []database.Record {
+	for i := range records {
+		records[i].Meta = model.SearchMeta{}
+	}
+	return records
+}
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/localstore/objectstore/queries_test.go:36]: nope, they need to be separated

File: pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		fragment.Ranges = append(fragment.Ranges, [2]int{termLocation.Start - f.Start, termLocation.End - f.Start})
+		// add the stuff before this location
+		// start the <mark> tag
+		curr = termLocation.End
+	}
+	// add any remaining text after the last token
+	fragment.Text = append(fragment.Text, f.Orig[curr:f.End]...)
+
+	b, _ := json.Marshal(fragment)
+	return string(b)
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go]: I feel it's more easy to read this way

File: core/object.go [hunk 5]
```diff
@@ -142,6 +143,56 @@ func (mw *Middleware) ObjectSearch(cctx context.Context, req *pb.RpcObjectSearch
...
+			Details:  rec.Details,
+			Meta:     []*model.SearchMeta{&(results[i].Meta)},
+		})
+	}
+	if err != nil {
+		return response(pb.RpcObjectSearchWithMetaResponseError_UNKNOWN_ERROR, nil, err)
+	}
+
+	return response(pb.RpcObjectSearchWithMetaResponseError_NULL, resultsModels, nil)
+}
```
[the remark this answers is not in the document]
Comment (requilence) [reply core/object.go:194]: ForceFTIndex doesn't have error arg. it simply send the msg to the chan to prioritize fulltext indexing

File: core/indexer/fulltext.go [hunk 6]
```diff
@@ -149,6 +156,10 @@ func (i *indexer) prepareSearchDocument(ctx context.Context, id string, processo
 					return true
 				}
 
+				if len(pbtypes.GetStringList(b.Model().GetFields(), text.DetailsKeyFieldName)) > 0 {
```
Comment (fat-fellow) [inline core/indexer/fulltext.go:159]: tests?

File: core/indexer/fulltext.go [hunk 7]
```diff
@@ -123,6 +124,12 @@ func (i *indexer) prepareSearchDocument(ctx context.Context, id string, processo
 			if val == "" {
 				continue
 			}
+			// skip readonly and hidden system relations
+			if bundledRel, err := bundle.PickRelation(domain.RelationKey(rel.Key)); err == nil {
```
Comment (fat-fellow) [inline core/indexer/fulltext.go:128]: tests?

File: core/object.go [hunk 8]
```diff
@@ -142,6 +143,56 @@ func (mw *Middleware) ObjectSearch(cctx context.Context, req *pb.RpcObjectSearch
 	return response(pb.RpcObjectSearchResponseError_NULL, records2, nil)
 }
 
+func (mw *Middleware) ObjectSearchWithMeta(cctx context.Context, req *pb.RpcObjectSearchWithMetaRequest) *pb.RpcObjectSearchWithMetaResponse {
+	response := func(code pb.RpcObjectSearchWithMetaResponseErrorCode, results []*model.SearchResult, err error) *pb.RpcObjectSearchWithMetaResponse {
```
Comment (fat-fellow) [inline core/object.go:147]: should be using modern technologies by @deff7 see ObjectGraph for example

File: pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
+//  Copyright (c) 2014 Couchbase, Inc.
```
Comment (fat-fellow) [inline pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go:1]: should we have tests for this or is it temporary?

File: pkg/lib/localstore/objectstore/fixture.go [hunk 10]
```diff
@@ -74,6 +75,11 @@ func (fx *StoreFixture) Init(a *app.App) (err error) {
 
 type TestObject map[domain.RelationKey]*types.Value
 
+type TestObjectWithMeta struct {
+	TestObject
+	model.SearchMeta
+}
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/localstore/objectstore/fixture.go:81]: unused, removed

File: core/object.go [hunk 11]
```diff
@@ -142,6 +143,56 @@ func (mw *Middleware) ObjectSearch(cctx context.Context, req *pb.RpcObjectSearch
...
+
+	if mw.applicationService.GetApp() == nil {
+		return response(pb.RpcObjectSearchWithMetaResponseError_BAD_INPUT, nil, fmt.Errorf("account must be started"))
+	}
+
+	if req.FullText != "" {
+		mw.applicationService.GetApp().MustComponent(indexer.CName).(indexer.Indexer).ForceFTIndex()
+	}
+
+	ds := mw.applicationService.GetApp().MustComponent(objectstore.CName).(objectstore.ObjectStore)
```
[the remark this answers is not in the document]
Comment (requilence) [reply core/object.go:164]: agree, but it was the same in the ObjectSearch method. This method is just a new version that is going to replace the ObjectSearch.

File: pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go (hunk 9 above)
Comment (requilence) [reply pkg/lib/localstore/ftsearch/jsonhighlighter/json/json.go:1]: this is temporary for the bleve we are going to remove. Also tested as part of queries_test.go

File: pkg/lib/localstore/objectstore/queries.go [hunk 12]
```diff
@@ -2,33 +2,98 @@
...
+	"github.com/anyproto/anytype-heart/pkg/lib/localstore/ftsearch"
+	jsonFormatter "github.com/anyproto/anytype-heart/pkg/lib/localstore/ftsearch/jsonhighlighter/json"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
 	"github.com/anyproto/anytype-heart/space/spacecore/typeprovider"
 	"github.com/anyproto/anytype-heart/util/pbtypes"
+	text2 "github.com/anyproto/anytype-heart/util/text"
+)
+
+const (
+	minFulltextScore = 0.05
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/localstore/objectstore/queries.go]: done

File: pkg/lib/localstore/objectstore/queries.go [hunk 13]
```diff
@@ -73,45 +235,114 @@
...
 }
 
-func (s *dsObjectStore) makeFTSQuery(text string, filters *database.Filters) (*database.Filters, error) {
-	if s.fts == nil {
-		return filters, fmt.Errorf("fullText search not configured")
+// jsonHighlightToRanges converts json highlight to runes ranges
+// input ranges are positions of bytes, the returned ranges are position of runes
+func jsonHighlightToRanges(s string) (text string, ranges []*model.Range) {
+	fragment, err := jsonFormatter.UnmarshalFromString(s)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (requilence) [reply pkg/lib/localstore/objectstore/queries.go:264]: done

**Pull request merged**